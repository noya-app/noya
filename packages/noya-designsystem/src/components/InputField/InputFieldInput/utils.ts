export const ignoredProps = new Set([
  'labelPosition',
  'labelSize',
  'hasLabel',
  'hasDropdown',
  'textAlign',
  'variant',
  'onNudge',
]);
