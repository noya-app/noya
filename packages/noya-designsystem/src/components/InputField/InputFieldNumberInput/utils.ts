export function parseNumber(value: string) {
  return value ? Number(value) : NaN;
}

export function handleNudge(e: {
  key: string;
  shiftKey: boolean;
  altKey: boolean;
}): number | undefined {
  let handled = false;
  let amount = 0;

  switch (e.key) {
    case 'ArrowUp':
      amount = 1;
      handled = true;
      break;
    case 'ArrowDown':
      amount = -1;
      handled = true;
      break;
  }

  if (!handled) return;

  if (e.shiftKey) {
    amount *= 10;
  } else if (e.altKey) {
    amount *= 0.1;
  }

  return handled ? amount : undefined;
}
