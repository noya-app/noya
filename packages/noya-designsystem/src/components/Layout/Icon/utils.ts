interface GlyphMap {
  [name: string]: number;
}

type Icon =
  | any
  | {
      properties: {
        name: string;
        code: number;
      };
    };

export interface SelectionFileMinimal {
  icons: Icon[];
}

export function createGlyphsMap(selectionFile: SelectionFileMinimal) {
  const ghlyps: GlyphMap = {};

  selectionFile.icons.forEach((icon) => {
    if ('properties' in icon) {
      ghlyps[icon.properties.name] = icon.properties.code.toString(16);
    }
  });

  return ghlyps;
}
