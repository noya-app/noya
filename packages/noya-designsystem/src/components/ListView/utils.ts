import type { ListRowMarginType } from './types';

export function getPositionMargin(marginType: ListRowMarginType) {
  return {
    top: marginType === 'top' || marginType === 'vertical' ? 8 : 0,
    bottom: marginType === 'bottom' || marginType === 'vertical' ? 8 : 0,
  };
}
