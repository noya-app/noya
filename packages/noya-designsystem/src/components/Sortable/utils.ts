import type { DropValidator, RelativeDropPosition } from './types';

export function validateDropIndicator(
  acceptsDrop: DropValidator,
  activeId: string,
  overId: string,
  offsetTop: number,
  elementTop: number,
  elementHeight: number,
): RelativeDropPosition | undefined {
  const acceptsDropInside = acceptsDrop(activeId, overId, 'inside');

  // If we're in the center of the element, prefer dropping inside
  if (
    offsetTop >= elementTop + elementHeight / 3 &&
    offsetTop <= elementTop + (elementHeight * 2) / 3 &&
    acceptsDropInside
  )
    return 'inside';

  // Are we over the top or bottom half of the element?
  const indicator =
    offsetTop < elementTop + elementHeight / 2 ? 'above' : 'below';

  // Drop above or below if possible, falling back to inside
  return acceptsDrop(activeId, overId, indicator)
    ? indicator
    : acceptsDropInside
    ? 'inside'
    : undefined;
}

export const defaultAcceptsDrop: DropValidator = (
  sourceId,
  destinationId,
  position,
) => {
  return position !== 'inside' && sourceId !== destinationId;
};
