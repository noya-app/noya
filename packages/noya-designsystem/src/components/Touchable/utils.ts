import { GestureResponderEvent } from 'react-native';

import { Point } from 'noya-geometry';
import type {
  TouchMeta,
  TouchPoint,
  TouchCallback,
  TouchableContextType,
} from './types';

// Time to consider single press a long press
export const LongPressThresholdMS = 370;
// delta distance of average distance to gesture centroid
// to consider it a pinch gesture
export const PinchThreshold = 5;
// delta distance of gesture centroid between calls
// to consider it a pan gesture
export const PanThreshold = 5;

export const initMeta = {
  points: [],
  avgDistance: 0,
  centroid: { x: 0, y: 0 },
};

export const initialHandlers: TouchableContextType = {
  onPress: [],
  onLongPress: [],
  onTouchStart: [],
  onTouchUpdate: [],
  onTouchEnd: [],
};

export const touchableEventNames = [
  'onPress',
  'onLongPress',
  'onTouchStart',
  'onTouchUpdate',
  'onTouchEnd',
];

export function mergeHandlers(
  parentHandlers: TouchCallback[],
  currentHandler?: TouchCallback,
): TouchCallback[] {
  const handlers: TouchCallback[] = [...parentHandlers];

  if (currentHandler) {
    handlers.push(currentHandler);
  }

  return handlers;
}

export function getPoint({ nativeEvent }: GestureResponderEvent) {
  if (nativeEvent.touches.length > 1) {
    const firstTouch = nativeEvent.touches[0];

    return {
      x: firstTouch.locationX,
      y: firstTouch.locationY,
    };
  }

  return {
    x: Math.round(nativeEvent.locationX),
    y: Math.round(nativeEvent.locationY),
  };
}

export function getDistance(p1: Point, p2: Point) {
  return Math.round(Math.sqrt((p2.x - p1.x) ** 2 + (p2.y - p1.y) ** 2));
}

export function getTouchMeta({
  nativeEvent,
}: GestureResponderEvent): TouchMeta {
  const centroid: Point = { x: 0, y: 0 };
  let avgDistance = 0;

  const points: TouchPoint[] = nativeEvent.touches.map((touch) => ({
    x: Math.round(touch.locationX),
    y: Math.round(touch.locationY),
    id: touch.identifier,
  }));

  points.forEach((point) => {
    centroid.x += point.x;
    centroid.y += point.y;
  });

  centroid.x /= points.length;
  centroid.y /= points.length;

  points.forEach((point) => {
    avgDistance += getDistance(point, centroid);
  });

  avgDistance /= points.length;

  if (!points.length) {
    return {
      points: [],
      centroid: { x: nativeEvent.locationX, y: nativeEvent.locationY },
      avgDistance: 0,
    };
  }

  return { points, centroid, avgDistance };
}
