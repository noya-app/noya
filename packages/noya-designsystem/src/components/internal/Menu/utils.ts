import { MenuItem, SEPARATOR_ITEM } from './constants';

function getKeyboardShortcuts<T extends string>(
  item: MenuItem<T>,
): [string, T][] {
  if (item === SEPARATOR_ITEM) return [];

  if (item.items) return item.items.flatMap(getKeyboardShortcuts);

  if (item.disabled || !item.value || !item.shortcut) return [];

  return [[item.shortcut, item.value]];
}

export function getKeyboardShortcutsForMenuItems<T extends string>(
  menuItems: MenuItem<T>[],
  onSelect: (type: T) => void,
): Record<string, () => void> {
  return Object.fromEntries(
    menuItems
      .flatMap(getKeyboardShortcuts)
      .map(([key, value]) => [key, () => onSelect(value)]),
  );
}
