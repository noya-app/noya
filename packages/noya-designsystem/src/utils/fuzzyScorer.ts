import { IItemScore, scoreFilePathFuzzy } from 'vscode-fuzzy-scorer';

export function fuzzyScore({ item, query }: { item: string; query: string }) {
  return scoreFilePathFuzzy({ path: item, query });
}

export type IScoredItem = IItemScore & {
  index: number;
};

export function fuzzyFilter({
  items,
  query,
}: {
  items: string[];
  query: string;
}) {
  const scoredItems = items.map(
    (text, index): IScoredItem => ({
      index,
      ...fuzzyScore({ item: text, query }),
    }),
  );

  if (!query) return scoredItems;

  return scoredItems
    .filter(({ score }) => score > 0)
    .sort((a, b) => b.score - a.score);
}

export type IToken =
  | { type: 'text'; text: string }
  | { type: 'match'; text: string };

export function fuzzyTokenize({
  item,
  itemScore,
}: {
  item: string;
  itemScore: IItemScore;
}) {
  const tokens: IToken[] = [];

  let lastMatchIndex = 0;

  for (const match of itemScore.labelMatch ?? []) {
    if (match.start > lastMatchIndex) {
      tokens.push({
        type: 'text',
        text: item.slice(lastMatchIndex, match.start),
      });
    }

    tokens.push({
      type: 'match',
      text: item.slice(match.start, match.end),
    });

    lastMatchIndex = match.end;
  }

  if (lastMatchIndex < item.length) {
    tokens.push({
      type: 'text',
      text: item.slice(lastMatchIndex),
    });
  }

  return tokens;
}
