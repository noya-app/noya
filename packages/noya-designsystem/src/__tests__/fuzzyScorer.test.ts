import { fuzzyFilter, fuzzyScore, fuzzyTokenize } from '../utils/fuzzyScorer';

test('prefix match', () => {
  const result = fuzzyScore({ item: 'font-bold', query: 'font' });

  expect(result.labelMatch).toEqual([{ start: 0, end: 4 }]);
});

test('suffix match', () => {
  const result = fuzzyScore({ item: 'font-bold', query: 'bold' });

  expect(result.labelMatch).toEqual([{ start: 5, end: 9 }]);
});

test('non-contiguous match', () => {
  const result = fuzzyScore({ item: 'font-bold', query: 'fb' });

  expect(result.labelMatch).toEqual([
    { start: 0, end: 1 },
    { start: 5, end: 6 },
  ]);
});

test('filter single match', () => {
  const result = fuzzyFilter({
    items: ['font-bold', 'font-italic', 'font-regular'],
    query: 'bold',
  });

  expect(result).toMatchSnapshot();
});

test('filter multiple matches', () => {
  const result = fuzzyFilter({
    items: ['font-bold', 'border'],
    query: 'b',
  });

  expect(result).toMatchSnapshot();
});

test('tokenizes prefix match', () => {
  const item = 'font-bold';
  const query = 'font';
  const result = fuzzyTokenize({
    item,
    itemScore: fuzzyScore({ item, query }),
  });

  expect(result).toEqual([
    { type: 'match', text: 'font' },
    { type: 'text', text: '-bold' },
  ]);
});

test('tokenizes suffix match', () => {
  const item = 'font-bold';
  const query = 'bold';
  const result = fuzzyTokenize({
    item,
    itemScore: fuzzyScore({ item, query }),
  });

  expect(result).toEqual([
    { type: 'text', text: 'font-' },
    { type: 'match', text: 'bold' },
  ]);
});

test('tokenizes non-contiguous match', () => {
  const item = 'font-bold';
  const query = 'fb';
  const result = fuzzyTokenize({
    item,
    itemScore: fuzzyScore({ item, query }),
  });

  expect(result).toEqual([
    { type: 'match', text: 'f' },
    { type: 'text', text: 'ont-' },
    { type: 'match', text: 'b' },
    { type: 'text', text: 'old' },
  ]);
});
