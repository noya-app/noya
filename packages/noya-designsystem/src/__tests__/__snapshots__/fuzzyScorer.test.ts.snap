// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter multiple matches 1`] = `
Array [
  Object {
    "index": 1,
    "labelMatch": Array [
      Object {
        "end": 1,
        "start": 0,
      },
    ],
    "score": 131072,
  },
  Object {
    "index": 0,
    "labelMatch": Array [
      Object {
        "end": 6,
        "start": 5,
      },
    ],
    "score": 32768,
  },
]
`;

exports[`filter single match 1`] = `
Array [
  Object {
    "index": 0,
    "labelMatch": Array [
      Object {
        "end": 9,
        "start": 5,
      },
    ],
    "score": 32768,
  },
]
`;
